### ISSUE TITLE
prompt_idx in gmp_prompt not work in stage 0?

### ISSUE THREAD

**Issue by Iranb** [Visual-AI/PromptCCD#8]
I tried to print `prompt_idx` and `batched_prompt` during the training process, but I noticed that the results of mean filtering based on GMM seem to be exactly the same across different epochs and samples. Have you encountered a similar issue during training?

![image](https://github.com/user-attachments/assets/984faa8f-f69f-4ecf-b19b-7d7b3b46f7c5)

**Comment by fcendra** [comment]
Hi there, 

I haven't encountered this issue before, but it's normal during the initial stages since the GMM model isn't fully set up yet. I'll look into it as soon as I have some time and get back to you.

**Comment by Iranb** [comment]
Thank you very much for your reply.  After servel attempts, when training reaches 130 epochs, the GMM still assigns exactly the same cluster centers to the samples in the current batch. I am also trying to find the cause of the issue. Thank you again for your response.

**Comment by fcendra** [comment]
Hi @Iranb, I tried to print out the `idx` variable which corresponds to the retrieved prompt index results as shown in this [line of code](https://github.com/Visual-AI/PromptCCD/blob/d13db371fe6a3990b93510e0f507734736648096/models/promptccd_utils/gmp_known_K.py#L75C9-L75C12). The results look fine in my side as can be seen below: 

![Screenshot 2025-01-07 at 6 08 08 PM](https://github.com/user-attachments/assets/0454e74a-dd5d-4173-8581-657d4ddf005f)

I'm not sure what went wrong with the program on your end, but you might want to check the version of your [pycave](https://pycave.borchero.com/index.html) library. My version is `3.2.1`. Another thing that might be helpful to consider is to check how good is the validation accuracy. Additionally, it could be useful to evaluate the validation accuracy. If the model has learned good features, then the GMM should be fitted with discriminative features as well. I hope this helps.

**Comment by Iranb** [comment]
Thank you for your response. I am using Pycave version `3.2.1` and PyTorch version `1.10`. I will continue to investigate the cause of the issue.

**Comment by Iranb** [comment]
I re-clone the promptccd repo and only change the dataset path in datautil,  then train and print prompt index results same as you did. The results show below:

![image](https://github.com/user-attachments/assets/ee50591b-7482-4cc2-838f-61b59cf4e147)

I think the issue may be caused by the different PyTorch version. I will conduct experiments using higher PyTorch versions.

Thank you for your nice work 😊！！

**Comment by Iranb** [comment]
Try to retrain the model using PyTorch 2.1.2 and record the changes in the GMM prediction results.

I cloned the code from github again and only change the data path.

I got the same issue that GMM prediction seems crashed.

The training log listed in the file below.

[torch2.1.2.log](https://github.com/user-attachments/files/18371832/torch2.1.2.log)


Are these results trained from the [GitHub](https://github.com/Visual-AI/PromptCCD) repository?

> Hi @Iranb, I tried to print out the `idx` variable which corresponds to the retrieved prompt index results as shown in this [line of code](https://github.com/Visual-AI/PromptCCD/blob/d13db371fe6a3990b93510e0f507734736648096/models/promptccd_utils/gmp_known_K.py#L75C9-L75C12). The results look fine in my side as can be seen below:
> 
> ![Screenshot 2025-01-07 at 6 08 08 PM](https://private-user-images.githubusercontent.com/55789008/400724387-0454e74a-dd5d-4173-8581-657d4ddf005f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0OTIwNDEsIm5iZiI6MTczNjQ5MTc0MSwicGF0aCI6Ii81NTc4OTAwOC80MDA3MjQzODctMDQ1NGU3NGEtZGQ1ZC00MTczLTg1ODEtNjU3ZDRkZGYwMDVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDA2NDkwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZiNzk2NmEyYmIyMDg4MzBlMzY5ZDlmMDFhNzM0MzY1NTZiYjA0YjI0MmQzZjg5YjQ2MjFhOGUyY2ZiYjAzMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Wd_ukniz9TKwZ1P9-cmC3O0X0elWfpeH9YTBV3JArdE)
> 
> I'm not sure what went wrong with the program on your end, but you might want to check the version of your [pycave](https://pycave.borchero.com/index.html) library. My version is `3.2.1`. Another thing that might be helpful to consider is to check how good is the validation accuracy. Additionally, it could be useful to evaluate the validation accuracy. If the model has learned good features, then the GMM should be fitted with discriminative features as well. I hope this helps.

**Comment by Iranb** [comment]
> Hi @Iranb, I tried to print out the `idx` variable which corresponds to the retrieved prompt index results as shown in this [line of code](https://github.com/Visual-AI/PromptCCD/blob/d13db371fe6a3990b93510e0f507734736648096/models/promptccd_utils/gmp_known_K.py#L75C9-L75C12). The results look fine in my side as can be seen below:
> 
> ![Screenshot 2025-01-07 at 6 08 08 PM](https://private-user-images.githubusercontent.com/55789008/400724387-0454e74a-dd5d-4173-8581-657d4ddf005f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0OTI2NDgsIm5iZiI6MTczNjQ5MjM0OCwicGF0aCI6Ii81NTc4OTAwOC80MDA3MjQzODctMDQ1NGU3NGEtZGQ1ZC00MTczLTg1ODEtNjU3ZDRkZGYwMDVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTEwVDA2NTkwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxYjcwZWNmZDE3MmY5ZGRlMWUxOWRhMGNhZDIyYzE2NGVkZGM4NDQ3MmY3NmQ1YmE4MzM2OTFmZWYzYmMzN2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yAbLJv2SeDBf_tOBlsx7RFuExARcxoOtQYZu3-YtMCQ)
> 
> I'm not sure what went wrong with the program on your end, but you might want to check the version of your [pycave](https://pycave.borchero.com/index.html) library. My version is `3.2.1`. Another thing that might be helpful to consider is to check how good is the validation accuracy. Additionally, it could be useful to evaluate the validation accuracy. If the model has learned good features, then the GMM should be fitted with discriminative features as well. I hope this helps.



I clone the code and run with torch version 2.1.2 again. 
The training log a list here

[torch2.1.2.log](https://github.com/user-attachments/files/18371879/torch2.1.2.log)

Are the results on your side trained with the code from the [GitHub](https://github.com/Visual-AI/PromptCCD) repository?

**Comment by fcendra** [comment]
Hi @Iranb, yes the results are obtained from this repository. The environment list I used to obtain the results is attached here: [requirements.txt](https://github.com/user-attachments/files/18371943/requirements.txt). 

By looking at your log file, it seems like the error is caused by the nan value contained in the model features (which are used to retrieve the GMM's components). Can you verify your model by checking the validation results as shown in below picture?

<img width="515" alt="Screenshot 2025-01-10 at 3 16 56 PM" src="https://github.com/user-attachments/assets/a37d327a-0aa3-43eb-b3df-2e2d91f530c5" />

Hope this helps!

**Comment by Iranb** [comment]
Trying to retrain for 200 epochs and test the model. Thank you for your help. I will synchronize the eval results to this issue after the training is completed. 😊

**Comment by Iranb** [comment]
Through the evaluation of the model, I found that the performance of the model remains unchanged (and even improves), while at the same time, I observed that the GMM centers still have the problems discussed previously.

The eval results:
![image](https://github.com/user-attachments/assets/8f710321-e532-4332-8104-79728bb6902c)

The training log:
[screenlog_bash.log.zip](https://github.com/user-attachments/files/18383710/screenlog_bash.log.zip)

**Comment by fcendra** [comment]
Hi @Iranb ,

Thank you for reporting this issue and for the ongoing discussion. We’ve reviewed the details and tried several approaches, but we haven’t yet identified the root cause..

Please rest assured that we are actively working on resolving this and will keep you updated with any progress or findings. We appreciate your patience and understanding as we work through this.

If you have any additional information or insights, please feel free to share them.

Best regards,
Fernando

**Comment by Iranb** [comment]
Thank you again for open-sourcing such excellent work. I will also investigate the causes of the problem. If there is any progress, I will update it here.
